docs/index.rst

```diff
@@ -1,260 +1,427 @@
+a file, return an empty configuration. Each variable will have a
+``FileOrigin``, and its key will be prefixed by its section name separated by
+*sep*.
+
+::
+
+  user_config_paths(*paths)
+
+Return a list of paths to user-specific configuration directories
+conventional for the current platform, with *paths* appended to each using
+``os.path.join``.
+
+OS X    | ``~/.config``, ``~/Library/Application Support``
+Unix    | ``$XDG_CONFIG_HOME``, ``~/.config``
+Windows | ``%HOME%\AppData\Roaming``
+
+::
+
+  sys_config_path(*paths)
+
+Return path to the system-wide configuration directory (shared by all users)
+conventional for the current platform, with *paths* appended using
+``os.path.join``.
+
+Unix, OS X | ``/etc``
+Windows    | ``%APPDATA%``
+
+::
+
+  module_dir(module_name)
+
+Return the path to the directory where the named module is found.
+
+::
+
+  from_env(prefix, sep="_")
+
+Return a configuration pulled from the environment. For each environment
+variable whose name has the given *prefix*, the configuration will have a
+variable whose name is everything after the *prefix* with underscores
+replaced by *sep* (see caveat at the end of this description), whose value is
+a string, and whose origin is an ``EnvironmentOrigin``.
+
+Environment overrides are handy for scripts that cannot create a
+configuration file (or cannot edit an existing one) and cannot change the
+command line before invoking your application.
+
+One notable caveat is that most shells have a limited character set for
+environment variable names. Often, the OS will support all non-null
+characters, but shells will only support letters, numbers, and underscores.
+In those cases, environment variables can be used to override only variables
+with "well behaved names". To help overcome this limitation, this source
+supports overriding variables with non-underscore separators using the *sep*
+argument.
+
+::
+
+  from_args(argv, prefix)
+
+Return a pair. The first half will be a configuration parsed from the command
+line long options in *argv* that have the given *prefix*. Each variable will
+have a ``CommandLineOrigin``. The second half will be the remaining unparsed
+command line arguments in the same order they appeared. Example::
+
+  (config, args) = confit.from_args(["a", "--eg-hello", "world", "b"], "eg")
+  config["hello"].get() # "world"
+  args # ["a", "b"]
+
+::
+
+  class confit.FileOrigin(filename, line, [column])
+
+::
+
+  class confit.EnvironmentOrigin(variable)
+
+::
+
+  class confit.CommandLineOrigin(option)
+
+
+Custom sources
+^^^^^^^^^^^^^^
+
+To extend this set with your own custom source, define a function that
+returns a configuration::
+
+  class confit.Configuration(config)
+
+*config* should be a mapping from variable names to (value, origin) pairs.
+For origins that do not fit the built-in types, consider defining your own
+origin class. If you do, it is best practice to define a pretty-printing
+``__str__`` method.
```

Comment:
Dumb question; sorry (this is answered above). It might be worth enumerating the interfaces of `Configuration` and `Variable` exhaustively somewhere, though:

* `config[key]`:  Return a `Variable`.
* `var.get(..)`
* `var.origin`